pkg/nsx/services/subnetport/builder.go

```diff
@@ -112,19 +110,29 @@ func (service *SubnetPortService) buildSubnetPort(obj interface{}, nsxSubnet *mo
 	}
 
 	if util.NSXSubnetDHCPEnabled(nsxSubnet) {
-		// DHCP was never implemented for SubnetPort. Subnet's DHCP config is the only place to identify if port has DHCP config.
-		allocateAddresses = "NONE"
-	} else if isIPPool {
-		// If MAC address from spec is not in NSX MAC Pool and we set allocateAddresses as BOTH,
-		// we will get the error `User defined address bindings are not allowed on LogicalPort
-		// InternalLogicalPort/{id} as its VIF AttachmentContext contain IP/MAC Pool {1}.`
-		allocateAddresses = "IP_POOL"
-	} else if nsxSubnet.AdvancedConfig == nil || nsxSubnet.AdvancedConfig.StaticIpAllocation == nil ||
-		nsxSubnet.AdvancedConfig.StaticIpAllocation.Enabled == nil || !*nsxSubnet.AdvancedConfig.StaticIpAllocation.Enabled {
-		// StaticIpAllocation is false by default. Set allocateAddresses to NONE if StaticIpAllocation is not set or disabled.
-		allocateAddresses = "NONE"
+		// Subnet with DHCPServer/DHCPRelay.
+		if inNSXMacPool != nil && *inNSXMacPool {
+			allocateAddresses = "MAC_POOL"
+		} else {
+			// DHCP was never implemented for SubnetPort. Subnet's DHCP config is the only place to identify if port has DHCP config.
+			allocateAddresses = "NONE"
+		}
 	} else {
-		allocateAddresses = "BOTH"
+		if util.NSXSubnetStaticIPAllocationEnabled(nsxSubnet) {
+			// Subnet with Static IPAM.
+			if inNSXMacPool != nil && !*inNSXMacPool {
+				allocateAddresses = "IP_POOL"
```

Comment:
for both IP and MAC specified, and MAC in NSX mac pool case, we should set allocate_addresses: BOTH